instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapper.java

```diff
@@ -0,0 +1,141 @@
+  static final AttributeKey<String> ATTR_THROWABLE_NAME = AttributeKey.stringKey("throwable.name");
+  static final AttributeKey<String> ATTR_THROWABLE_MESSAGE =
+      AttributeKey.stringKey("throwable.message");
+  static final AttributeKey<List<String>> ATTR_NDC = AttributeKey.stringArrayKey("ndc");
+  static final AttributeKey<String> ATTR_FQCN = AttributeKey.stringKey("fqcn");
+  static final AttributeKey<String> ATTR_MARKER = AttributeKey.stringKey("marker");
+
+  private static final Map<Level, Severity> LEVEL_SEVERITY_MAP;
+
+  static {
+    Map<Level, Severity> levelSeverityMap = new HashMap<>();
+    levelSeverityMap.put(Level.ALL, Severity.TRACE);
+    levelSeverityMap.put(Level.TRACE, Severity.TRACE2);
+    levelSeverityMap.put(Level.DEBUG, Severity.DEBUG);
+    levelSeverityMap.put(Level.INFO, Severity.INFO);
+    levelSeverityMap.put(Level.WARN, Severity.WARN);
+    levelSeverityMap.put(Level.ERROR, Severity.ERROR);
+    levelSeverityMap.put(Level.FATAL, Severity.FATAL);
+    LEVEL_SEVERITY_MAP = Collections.unmodifiableMap(levelSeverityMap);
+  }
+
+  static void mapLogEvent(LogBuilder builder, LogEvent logEvent) {
+    AttributesBuilder attributes = Attributes.builder();
+
+    // message
+    Message message = logEvent.getMessage();
+    if (message != null) {
+      builder.setBody(message.getFormattedMessage());
+    }
+
+    // time
+    builder.setEpoch(logEvent.getNanoTime(), TimeUnit.NANOSECONDS);
+
+    // level
+    Level level = logEvent.getLevel();
+    if (level != null) {
+      builder.setSeverity(
+          LEVEL_SEVERITY_MAP.getOrDefault(level, Severity.UNDEFINED_SEVERITY_NUMBER));
+      builder.setSeverityText(logEvent.getLevel().name());
+    }
+
+    // logger
+    attributes.put(ATTR_LOGGER_NAME, logEvent.getLoggerName());
+
+    // fully qualified class name
+    attributes.put(ATTR_FQCN, logEvent.getLoggerFqcn());
+
+    // thread
+    attributes.put(ATTR_THREAD_NAME, logEvent.getThreadName());
+    attributes.put(ATTR_THREAD_ID, logEvent.getThreadId());
+    attributes.put(ATTR_THREAD_PRIORITY, logEvent.getThreadPriority());
+
+    // throwable
+    Throwable throwable = logEvent.getThrown();
+    if (throwable != null) {
+      attributes.put(ATTR_THROWABLE_NAME, throwable.getClass().getName());
+      attributes.put(ATTR_THROWABLE_MESSAGE, throwable.getMessage());
+    }
+
+    // marker
+    Marker marker = logEvent.getMarker();
+    if (marker != null) {
+      attributes.put(ATTR_MARKER, marker.getName());
+    }
+
+    // nested diagnostic context
+    List<String> contextStackList = logEvent.getContextStack().asList();
+    if (contextStackList.size() > 0) {
+      attributes.put(ATTR_NDC, contextStackList);
+    }
+
+    // mapped diagnostic context (included span context)
+    ReadOnlyStringMap contextData = logEvent.getContextData();
+    if (contextData != null) {
+      Map<String, String> contextMap = contextData.toMap();
+      if (contextMap != null) {
+        String traceId =
```

Comment:
I would like to first try with `Context.current()` and then consider alternatives if we find, e.g. people really want to wrap OTel in a async appenders, it is a significant usability problem.

Reply:
Sounds like a good plan.